### PR TITLE
doc: RuntimeAdapter

### PR DESCRIPTION
**Pull request by magne4000**

### CODE REVIEW

**Comment by magne4000** [comment]
@brillout regarding our discussion https://github.com/vikejs/vike-node/issues/26#issuecomment-2531082291

I published some doc around `RuntimeAdapter` that I merged for convenience https://universal-middleware.dev/reference/runtime-adapter

Do you think it's enough and understandable if we link this page from `vike-node` README?

**Comment by brillout** [comment]
LGTM, awesome. I'll leave a handful of comments about polishing potential.

File: docs/reference/runtime-adapter.md [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+// identify which adapter is currently used
+runtime.adapter;
+//      ^^^^^^^
+// runtime environment identifier. See https://runtime-keys.proposal.wintercg.org/
+runtime.runtime;
+//      ^^^^^^^
+// extracted route parameters. See http://localhost:5173/recipes/params-handler
+runtime.params;
+//      ^^^^^^
+// original `req` if running inside node
```
Comment (brillout) [inline docs/reference/runtime-adapter.md:242]: The original Node.js HTTP request object, see https://nodejs.org/api/http.html#class-httpincomingmessage

File: docs/reference/runtime-adapter.md [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+runtime.req;
+//      ^^^
+// original `res` if running inside node
```
Comment (brillout) [inline docs/reference/runtime-adapter.md:245]: The original Node.js HTTP response object (if available), see https://nodejs.org/api/...

File: docs/reference/runtime-adapter.md [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+type RuntimeAdapter = Explain<_RuntimeAdapter>;
+
+// ---cut---
+const runtime: RuntimeAdapter;
+
+// identify which adapter is currently used
+runtime.adapter;
+//      ^^^^^^^
+// runtime environment identifier. See https://runtime-keys.proposal.wintercg.org/
```
Comment (brillout) [inline docs/reference/runtime-adapter.md:236]: The runtime name ('hono', 'express', ...). Follows https://runtime-keys.proposal.wintercg.org/

File: docs/reference/runtime-adapter.md [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+  req?: IncomingMessage;
+  res?: ServerResponse;
+}
+
+type RuntimeAdapter = Explain<_RuntimeAdapter>;
+
+// ---cut---
+const runtime: RuntimeAdapter;
+
+// identify which adapter is currently used
```
Comment (brillout) [inline docs/reference/runtime-adapter.md:233]: // Identify which adapter is currently used. For example 'cloudlfare-workers' when using @universal-middleware/cloudlare-workers

File: docs/reference/runtime-adapter.md [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+
+type RuntimeAdapter = Explain<ExpressAdapter>;
+
+// ---cut---
+import type { RuntimeAdapter } from "somelib/express";
+
+const runtime: RuntimeAdapter;
+
+// original express `req` and `res`. See https://expressjs.com/en/guide/using-middleware.html
+runtime.express;
```
Comment (brillout) [inline docs/reference/runtime-adapter.md:137]: // original express `req` object, see https://expressjs.com/en/guide/using-middleware.html
runtime.express.req;
// original express  `res` object. See https://expressjs.com/en/guide/using-middleware.html
runtime.express.res;

File: docs/reference/runtime-adapter.md [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+
+type RuntimeAdapter = Explain<FastifyAdapter>;
+
+// ---cut---
+import type { RuntimeAdapter } from "somelib/fastify";
+
+const runtime: RuntimeAdapter;
+
+// original fastify `request` and `reply`. See https://fastify.dev/docs/latest/Guides/Getting-Started/
+runtime.fastify;
```
Comment (brillout) [inline docs/reference/runtime-adapter.md:158]: // original fastify `request` object ...
runtime.fastify;
// original fastify `reply` ...
runtime.fastify;

File: docs/reference/runtime-adapter.md [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+import type { RuntimeAdapter } from "somelib/hono";
+
+const runtime: RuntimeAdapter;
+
+// original Hono Context. See https://hono.dev/docs/api/context
+runtime.hono;
+//      ^^^^
+```
+
+```ts twoslash [h3]
```
Comment (brillout) [inline docs/reference/runtime-adapter.md:32]: Would be nice if the tab elements appears in order of popularity / what we like most. E.g. `express` first, then `hono`, then cloudflare, etc.

Comment (brillout) [reply docs/reference/runtime-adapter.md:32]: (Or `hono` first we feel like it.)

**Comment by brillout** [comment]
Little thing in case it's easy to change: the inline code blocks have a blue text so I thought there were links, but there aren't links.

**Comment by magne4000** [comment]
Thanks for the review :clap: 

> Little thing in case it's easy to change: the inline code blocks have a blue text so I thought there were links, but there aren't links.

Now that you mention it, it feels weird. It should be easy to change

File: docs/reference/runtime-adapter.md (hunk 7 above)
Comment (magne4000) [reply docs/reference/runtime-adapter.md:32]: https://github.com/magne4000/universal-middleware/commit/58196240017cf75472cadf65017128643f53c151

**Pull request merged**